docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook.go

```diff
@@ -0,0 +1,182 @@
+*/
+
+var log = logf.Log.WithName("cronjob-resource")
+
+/*
+Notice that we use kubebuilder markers to generate some webhook manifests.
+The first and second markers are responsible for generating a mutating webhook manifest and a validating
+webhook manifest respectively.
+
+The meaning of each marker can be found [here](../TODO.md).
+*/
+
+// +kubebuilder:webhook:path=/mutate-batch-tutorial-kubebuilder-io-v1-cronjob,mutating=true,failurePolicy=fail,groups=batch.tutorial.kubebuilder.io,resources=cronjobs,verbs=create;update,versions=v1,name=mcronjob.kb.io
+// +kubebuilder:webhook:path=/validate-batch-tutorial-kubebuilder-io-v1-cronjob,mutating=false,failurePolicy=fail,groups=batch.tutorial.kubebuilder.io,resources=cronjobs,verbs=create;update,versions=v1,name=vcronjob.kb.io
+
+/*
+We implement the `webhook.Defaulter` interface for our CronJob CRD.
+`webhook.Defaulter` only has one method: `Default()`.
+If `webhook.Defaulter` is implemented, a webhook server will be autowired to
+serve as the mutating webhook for CronJobs.
+*/
+
+var _ webhook.Defaulter = &CronJob{}
+
+// Default implements webhookutil.defaulter so a webhook will be registered for the type
+func (c *CronJob) Default() {
+	log.Info("defaulting cronjob", "namespace", c.Namespace, "name", c.Name)
+
+	if c.Spec.ConcurrencyPolicy == "" {
+		c.Spec.ConcurrencyPolicy = AllowConcurrent
+	}
+	if c.Spec.Suspend == nil {
+		c.Spec.Suspend = new(bool)
+	}
+	if c.Spec.SuccessfulJobsHistoryLimit == nil {
+		c.Spec.SuccessfulJobsHistoryLimit = new(int32)
+		*c.Spec.SuccessfulJobsHistoryLimit = 3
+	}
+	if c.Spec.FailedJobsHistoryLimit == nil {
+		c.Spec.FailedJobsHistoryLimit = new(int32)
+		*c.Spec.FailedJobsHistoryLimit = 1
+	}
+}
+
+/*
+We implement the `webhook.Validator` interface for our CronJob CRD.
+`webhook.Validator` interface has two methods: `ValidateCreate()` and `ValidateUpdate()`.
+If `webhook.Validator` is implemented, a webhook server will be autowired to
+serve as the validating webhook for CronJobs.
+*/
+
+var _ webhook.Validator = &CronJob{}
+
+// ValidateCreate implements webhookutil.validator so a webhook will be registered for the type
+func (c *CronJob) ValidateCreate() error {
+	log.Info("validate create", "namespace", c.Namespace, "name", c.Name)
+	return c.validateCronJob()
+}
+
+// ValidateUpdate implements webhookutil.validator so a webhook will be registered for the type
+func (c *CronJob) ValidateUpdate(old runtime.Object) error {
+	log.Info("validate update", "namespace", c.Namespace, "name", c.Name)
+	return c.validateCronJob()
+}
+
+/*
+We uses the same the validation logic for creation and update.
+*/
+
+func (c *CronJob) validateCronJob() error {
+	allErrs := c.validateCronJobSpec()
+	if len(c.ObjectMeta.Name) > apimachineryutilvalidation.DNS1035LabelMaxLength-11 {
+		// The cronjob controller appends a 11-character suffix to the cronjob (`-$TIMESTAMP`) when
+		// creating a job. The job name length limit is 63 characters.
```

Comment:
say why -- `the job name length is 63 character like all Kubernetes objects (which must fit in a DNS subdomain)`

Reply:
Add a note that we could validate this declaratively, but we keep it here as an example

Reply:
> Add a note that we could validate this declaratively

This is `ObjectMeta.Name` which is not owned by the user. Not sure if it's doable through validation schema. Thoughts?

Reply:
it's explicitly doable, in fact.  `GenerateName` and `Name` are the only things that are allowed to have validation, going forward.  Unfortunately, we don't have a good way to accomplish it in KB, but it's doable in in Kubernetes.

Reply:
Yes, that's what I meant.
Upstream k/k can add a marker for `ObjectMeta.Name` to have maximum 63 length. But as a downstream user, if I can't change the marker in upstream to accomplish my need which is 52 max length.